### PR TITLE
CB-11534 This plugin should support Android Marshmallow runtime permission

### PR DESCRIPTION
**Pull request by homen**
### Platforms affected

Android
### What does this PR do?

It will support Android Marshmallow runtime permission.
### What testing has been done on this change?

Tested in Android.
### Checklist
- [x] [ICLA](http://www.apache.org/licenses/icla.txt) has been signed and submitted to secretary@apache.org.
- [x] [Reported an issue](http://cordova.apache.org/contribute/issues.html) in the JIRA database
- [x] Commit message follows the format: "CB-3232: (android) Fix bug with resolving file paths", where CB-xxxx is the JIRA ID & "android" is the platform affected.
- [ ] Added automated test coverage as appropriate for this change.

### CODE REVIEW

**Comment by cordova-qa** [comment]
Cordova CI Build has one or more failures. 

**Commit**     - [Link](https://github.com/apache/cordova-plugin-file-transfer/pull/150/commits/1e0d0e02667c1a5e8f7eb20520d737a3947fa9d3)
**Dashboard** - [Link](http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/31/)

| Builder Name  | Console Output | Test Report | Device Logs  |
|     :---:     |     :---:      |   :---:     |     :---:    |
| [Windows 8.1 Store]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/31//PLATFORM=windows-8.1-store/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/31//PLATFORM=windows-8.1-store/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/31//PLATFORM=windows-8.1-store/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/31//PLATFORM=windows-8.1-store/artifact/) |
| [Windows 10  Store]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/31//PLATFORM=windows-10-store/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/31//PLATFORM=windows-10-store/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/31//PLATFORM=windows-10-store/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/31//PLATFORM=windows-10-store/artifact/) |
| [Windows 8.1 Phone]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/31//PLATFORM=windows-8.1-phone/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/31//PLATFORM=windows-8.1-phone/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/31//PLATFORM=windows-8.1-phone/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/31//PLATFORM=windows-8.1-phone/artifact/) |
| [iOS]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/31//PLATFORM=ios/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/31//PLATFORM=ios/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/31//PLATFORM=ios/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/31//PLATFORM=ios/artifact/) |
| [Android]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/31//PLATFORM=android/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/31//PLATFORM=android/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/31//PLATFORM=android/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/31//PLATFORM=android/artifact/) |

**Comment by homen** [comment]
What is the error?

**Comment by vladimir-kotikov** [comment]
@homen, there is nothing wrong with this PR (at least from CI perspective) - the failure was caused by some test defects on iOS platform. We're aware of that issue and investigating it.

**Comment by homen** [comment]
Ok @vladimir-kotikov

**Comment by huhaixiang** [comment]
When to support the Android 6 platform? Thanks

**Comment by homen** [comment]
@huhaixiang It is about Runtime  Permission.You can check here [link](https://developer.android.com/training/permissions/requesting.html)

**Comment by infil00p** [comment]
Hey.  This should really add the cordova-compat plugin if you're going to start adding permissions to plugins.

**Comment by homen** [comment]
@infil00p I do'nt understand?What I need to do?

**Comment by naveedahmed1** [comment]
Can this be merged? So that it can be used on Android 6 or higher.

**Comment by maxpaj** [comment]
Any progress on this?

**Comment by luckylooke** [comment]
Same as @maxpaj I can confirm this is a fix for permission issues on Android 6+, is there any reason this is not merged yet??

**Comment by maxpaj** [comment]
This should really be merged...

**Comment by ygrishajev** [comment]
@homen you should probably import `cordova.PermissionHelper` like [here](https://github.com/apache/cordova-plugin-camera/commit/a9c18710f23e86f5b7f8918dfab7c87a85064870#diff-f2e83abdaf34884df7e74b6493cf2dd7R37) and set a dependency for `cordova-plugin-compat` like [here](https://github.com/apache/cordova-plugin-camera/commit/a9c18710f23e86f5b7f8918dfab7c87a85064870#diff-53f390d375398624afe1cfe1125f42bfR33)

**Comment by stalniy** [comment]
@homen  do you plan to fix the PR in the nearest future? :) 
It's really annoying for me... And I want to have this fix in my codebase!

**Comment by homen** [comment]
Hi all,
I have done the required changes. It can be merged now.

**Comment by cordova-qa** [comment]
Cordova CI Build has one or more failures. 

**Commit**     - [Link](https://github.com/apache/cordova-plugin-file-transfer/pull/150/commits/640bd7150ebc994ea0398ee186be3b334b006080)
**Dashboard** - [Link](http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/82/)

 306 tests run, 15 skipped, 148 failed.

File: plugin.xml [hunk 1]
```diff
@@ -30,6 +30,8 @@
     <issue>https://issues.apache.org/jira/browse/CB/component/12320650</issue>
 
     <dependency id="cordova-plugin-file" version="^4.0.0" />
+	
+	<dependency id="cordova-plugin-compat" version="^1.0.0" />
```
Comment (filmaj) [inline plugin.xml]: Can you please use spaces instead of tabs, to keep it consistent?

**Comment by homen** [comment]
@filmaj All my editors showing correct alignment. Anyhow I fixed this with five commits :)

**Comment by cordova-qa** [comment]
Cordova CI Build has one or more failures. 

**Commit**     - [Link](https://github.com/apache/cordova-plugin-file-transfer/pull/150/commits/4be0f089431427aad2aa6c42146d45c31fdcb6b6)
**Dashboard** - [Link](http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/83/)

 306 tests run, 15 skipped, 149 failed.

**Comment by cordova-qa** [comment]
Cordova CI Build has one or more failures. 

**Commit**     - [Link](https://github.com/apache/cordova-plugin-file-transfer/pull/150/commits/f4669ec2f5b20dfeb9aaba3cc3bf4c49cf553abf)
**Dashboard** - [Link](http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/84/)

 306 tests run, 15 skipped, 148 failed.

**Comment by cordova-qa** [comment]
Cordova CI Build has one or more failures. 

**Commit**     - [Link](https://github.com/apache/cordova-plugin-file-transfer/pull/150/commits/eaabd8e84e08ef0090d67f32b504d1ff20c75f9d)
**Dashboard** - [Link](http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-file-transfer-pr/85/)

 306 tests run, 15 skipped, 153 failed.

**Comment by filmaj** [comment]
Thanks. I will need to test this manually as the file-transfer CI is super broken.

**Comment by stalniy** [comment]
@filmaj any news on this?

**Comment by filmaj** [comment]
Alright, I took a look at this again after an extended break and I have more comments, and want to bring more information to the community about the future of this plugin as well.

First, and I think most important, we need to be clear about the future of this plugin. A couple of months back, Cordova developers did an [audit of the core cordova plugins](https://issues.apache.org/jira/browse/CB-12708), and one result was the decision to [sunset the File Transfer plugin](https://issues.apache.org/jira/browse/CB-12711) in [favour of the increasingly-available XHR2 APIs](http://caniuse.com/#feat=xhr2). The XHR2 (now simply merged into the core XHR spec) features that are relevant to File Transfer plugin users are the ability to upload and to receive progress events - giving us exact feature parity for what File Transfer provides to users today. The benefit to sunsetting this plugin is that there would be less stuff to maintain for Cordova as XHR2 is already widely available on most mobile platform. With that in mind, that makes me reluctant to make many changes to this plugin as it's essentially makework.

Second, looking at the specific code changes in here, I started wondering, why execute the permission requesting inside this plugin? This plugin already depends on the core cordova File plugin, which does have support for Marshmallow permission model - why duplicate that? Turns out it is because the read/write methods of File Transfer, on the native Android side, are completely recreated/reproduced! I think a "proper" change to FT to support this would be to leverage [cordova-plugin-file's `FileUtils` class and call its `execute` method, passing in the right arguments to trigger the `write` action](https://github.com/apache/cordova-plugin-file/blob/master/src/android/FileUtils.java#L351-L369). Similarly, for reading files, the same could be done, using one of the [plethora of `readAs*` methods available in cordova-plugin-file's `FileUtils.java` class](https://github.com/apache/cordova-plugin-file/blob/master/src/android/FileUtils.java#L310-L349). The benefits I see to using this approach:
1. no need to rewrite permission handling code - cordova-plugin-file's native Android code already does that.
2. we could completely remove all file-reading and file-writing native Android code inside cordova-plugin-file-transfer, as we'd defer all of that work to the File plugin.

All that being said, once Cordova sunsets this plugin, the plugin code will be open and available to the community. The community could manage this code as it is however it seems right to them. While the patch here is not "perfect", it works, and that is more valuable to a lot of people.

So, I think I won't take any action on this PR. I am sorry that my previous review and comments in here made it seem like this PR would make its way into the mainline. Things have changed, and this plugin's window of life is closing.

**Comment by maverickmishra** [comment]
With the new features introduced in [XMLHttpRequest](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest), this plugin is not needed any more. Migrating from this plugin to using the new features of [XMLHttpRequest](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest), is explained in this [Cordova blog post](https://cordova.apache.org/blog/2017/10/18/from-filetransfer-to-xhr2.html).

**Pull request closed without merging**